connect/proxy/listener_test.go

```diff
@@ -19,71 +19,99 @@ import (
 	"github.com/hashicorp/consul/lib/freeport"
 )
 
-func testSetupMetrics(t *testing.T) *metrics.InmemSink {
+type MockSink struct {
+	keys   [][]string
+	vals   []float32
+	labels [][]metrics.Label
+}
+
+func (m *MockSink) SetGaugeWithLabels(key []string, val float32, labels []metrics.Label) {
+	m.keys = append(m.keys, key)
+	m.vals = append(m.vals, val)
+	m.labels = append(m.labels, labels)
+}
+
+func (m *MockSink) IncrCounterWithLabels(key []string, val float32, labels []metrics.Label) {
+	m.keys = append(m.keys, key)
+	m.vals = append(m.vals, val)
+	m.labels = append(m.labels, labels)
+}
+
+// Empty methods below only included to satisfy MetricSink interface
+func (m *MockSink) EmitKey(key []string, val float32) {
+}
+func (m *MockSink) SetGauge(key []string, val float32) {
+}
+func (m *MockSink) IncrCounter(key []string, val float32) {
+}
+func (m *MockSink) AddSample(key []string, val float32) {
+}
+func (m *MockSink) AddSampleWithLabels(key []string, val float32, labels []metrics.Label) {
+}
+
+// Flattens the key for formatting along with its labels, removes spaces
+func (m *MockSink) flattenKeyLabels(parts []string, labels []metrics.Label) (string, string) {
+	buf := &bytes.Buffer{}
+	replacer := strings.NewReplacer(" ", "_")
+
+	if len(parts) > 0 {
+		replacer.WriteString(buf, parts[0])
+	}
+	for _, part := range parts[1:] {
+		replacer.WriteString(buf, ".")
+		replacer.WriteString(buf, part)
+	}
+
+	key := buf.String()
+
+	for _, label := range labels {
+		replacer.WriteString(buf, fmt.Sprintf(";%s=%s", label.Name, label.Value))
+	}
+
+	return buf.String(), key
+}
+
+func testSetupMetrics(t *testing.T) *MockSink {
 	// Record for ages (5 mins) so we can be confident that our assertions won't
 	// fail on silly long test runs due to dropped data.
-	s := metrics.NewInmemSink(10*time.Second, 300*time.Second)
+	s := MockSink{}
 	cfg := metrics.DefaultConfig("consul.proxy.test")
 	cfg.EnableHostname = false
-	metrics.NewGlobal(cfg, s)
-	return s
+	metrics.NewGlobal(cfg, &s)
+	return &s
 }
 
-func assertCurrentGaugeValue(t *testing.T, sink *metrics.InmemSink,
+func assertTelemetryValue(t *testing.T, sink *MockSink,
```

Comment:
Hmm loosing type seems to be too lossy here - it's important if we want a counter that it actually is a counter not a gauge by mistake etc.

I suggest this whole method could be rewritten into methods of MockSink e.g. `GetCounterValue(name string) float64` and then the assertions can just be made normally in the test body. The only reason this was a helper before was that it was _soo_ much work to prise open the InMemSink internals to get the value.

----

Wait, later I realised this is a limitation of the `MockSink` currently. As noted in main comment I'd like to change that and I don't think it's super hard, but see what you think and we can figure out if it's worth making this as an interim change now or just doing what I suggest (or not doing that at all if others disagree!)